crowbar_framework/app/models/pacemaker_service.rb

```diff
@@ -75,6 +75,151 @@ def create_proposal
     base
   end
 
+  # Small helper to expand all items (nodes, clusters) used inside an applied
+  # proposal
+  def expand_nodes_in_barclamp_role(cluster_role, node_object_all)
+    all_nodes_for_cluster_role = cluster_role.elements.values.flatten.compact.uniq
+
+    all_nodes_for_cluster_role_expanded, failures = expand_nodes_for_all(all_nodes_for_cluster_role)
+    unless failures.nil? || failures.empty?
+      @logger.debug "[pacemaker] expand_nodes_in_barclamp_role: skipping items that we failed to expand: #{failures.join(", ")}"
+    end
+
+    # Do not keep deleted nodes
+    all_nodes_for_cluster_role_expanded = all_nodes_for_cluster_role_expanded & node_object_all.map{|n| n.name}
+
+    all_nodes_for_cluster_role_expanded
+  end
+
+  def apply_cluster_roles_to_new_nodes(role, member_nodes)
+    ### Beware of possible confusion between different level of "roles"!
+    # - we have barclamp roles that are related to a barclamp (as in "knife role
+    #   list | grep config" or RoleObject.proposal?); the cluster_role variable
+    #   is always such a role
+    # - we have roles inside each barclamp roles (as in "the role I assign to
+    #   nodes, like provisioner-server")
+
+    # Make sure that all nodes in the cluster have all the roles assigned to
+    # this cluster.
+    required_barclamp_roles = []
+    required_pre_chef_calls = []
+    cluster_element = "#{PacemakerServiceObject.cluster_key}:#{role.inst}"
+
+    # Find all barclamp roles where this cluster is used
+    cluster_roles = RoleObject.all.select do |role_object|
+      role_object.proposal? && \
+      role_object.elements.values.flatten.compact.uniq.include?(cluster_element)
+    end
+
+    # Inside each barclamp role, identify which role is required
+    for cluster_role in cluster_roles do
+      service = ServiceObject.get_service(cluster_role.barclamp).new(Rails.logger)
+      runlist_priority_map = cluster_role.override_attributes["element_run_list_order"] || {}
+      role_map = cluster_role.override_attributes["element_states"] || {}
+
+      deployment = cluster_role.override_attributes[cluster_role.barclamp]
+      save_it = false
+
+      cluster_role.elements.each do |role_name, node_names|
+        next unless node_names.include?(cluster_element)
+
+        required_barclamp_roles << { :service => service,
+                                     :barclamp_role => cluster_role,
+                                     :name => role_name,
+                                     :priority => runlist_priority_map[role_name] || service.chef_order,
+                                     :element_states => role_map[role_name] }
+
+        # Update elements_expanded attribute
+        expanded_nodes, failures = expand_nodes_for_all(node_names)
+        unless failures.nil? || failures.empty?
+          @logger.debug "[pacemaker] apply_cluster_roles_to_new_nodes: skipping items that we failed to expand: #{failures.join(", ")}"
```

Comment:
@aspiers this could happen if a proposal involves more than one cluster, and one cluster actually got unapplied in the meantime. To be honest, it's something that we should make impossible in the first place (but that's totally unrelated to this PR).

We're just not failing here because if we fail to apply now, then we won't run the pre_chef bits for the new members of the cluster, which is really bad.

I'll change the log level to a warning.